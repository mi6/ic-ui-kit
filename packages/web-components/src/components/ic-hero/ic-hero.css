@import "../../global/normalize.css";

:host {
  display: block;
  background-color: var(--ic-theme-tertiary);
  color: var(--ic-theme-text);

  --hero-heading-bottom-margin: var(--ic-space-md);
}

:host(.dark) {
  background-color: var(--ic-theme-secondary);
}

:host(.has-background-image) {
  background-repeat: no-repeat;
  background-position: right -6.25rem;
  background-size: auto calc(100% + 6.25rem);
  box-shadow: var(--ic-elevation-inset);
}

@media (prefers-reduced-motion) {
  :host(.has-background-image) {
    background-position: right -6.25rem !important;
  }
}

ic-typography.heading-bottom-spacing {
  margin-bottom: var(--hero-heading-bottom-margin);
}

.section-container {
  display: flex;
  align-items: center;
  height: 100%;
}

.left-container {
  padding: var(--ic-space-xl) 0;
}

.left-container:not(.left-container-full-width) {
  flex-basis: 50%;
}

.left-container-full-width {
  width: 100%;
}

.right-container {
  flex-basis: 50%;
}

.interaction-container {
  display: flex;
  gap: var(--ic-space-md);
  margin-top: var(--ic-space-lg);
  flex-wrap: wrap;
}

.secondary-container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  border-left: var(--ic-space-xxxs) solid var(--ic-theme-text);
}

.secondary-subheading {
  margin-top: var(--ic-space-xxs);
}

:host([content-aligned="left"]) .left-container,
:host([content-aligned="left"]) .interaction-container {
  justify-content: flex-start;
  text-align: left;
}

:host([content-aligned="center"]) .left-container,
:host([content-aligned="center"]) .interaction-container {
  justify-content: center;
  text-align: center;
}

@media (min-width: 992px) {
  .left-container-full-width {
    width: 66.6%;
  }
}

@media (min-width: 1044px) {
  :host,
  .section-container {
    min-height: 13rem;
  }

  :host(.small),
  :host(.small) .section-container {
    min-height: 10.813rem;
  }

  .left-container:not(.left-container-full-width) {
    flex-basis: 66.6%;
  }

  .left-container {
    min-height: 9rem;
  }

  .right-container {
    flex-basis: 33.3%;
    margin-left: 3.125rem;
  }

  .secondary-container {
    min-height: 9rem;
  }

  .secondary-heading,
  .secondary-subheading {
    margin-left: 3.938rem;
  }
}

@media (min-width: 801px) and (max-width: 1043px) {
  :host(.has-background-image) {
    background-image: none !important;
  }

  :host,
  .section-container {
    min-height: 16rem;
  }

  :host(.small),
  :host(.small) .section-container {
    min-height: 15rem;
  }

  :host(.secondary-heading),
  :host(.secondary-heading) .section-container {
    min-height: 14rem;
  }

  .left-container {
    min-height: 10rem;
  }

  .secondary-container {
    min-height: 10rem;
  }

  .secondary-heading,
  .secondary-subheading {
    margin-left: 5.922rem;
  }

  .right-container {
    margin-left: 3.125rem;
  }
}

@media (min-width: 641px) and (max-width: 800px) {
  :host(.has-background-image) {
    background-image: none !important;
  }

  :host,
  .section-container {
    min-height: 16rem;
  }

  :host(.small),
  :host(.small) .section-container {
    min-height: 15rem;
  }

  :host(.secondary-heading),
  :host(.secondary-heading) .section-container {
    min-height: 14rem;
  }

  .left-container {
    min-height: 10rem;
  }

  .secondary-container {
    min-height: 10rem;
  }

  .secondary-heading,
  .secondary-subheading {
    margin-left: 2.125rem;
  }

  .right-container {
    margin-left: 3.125rem;
  }
}

@media (min-width: 481px) and (max-width: 640px) {
  :host(.has-background-image) {
    background-image: none !important;
  }

  :host,
  .section-container {
    min-height: 13rem;
  }

  :host(.small),
  :host(.small) .section-container {
    min-height: 10.813rem;
  }

  :host(.secondary-heading),
  :host(.secondary-heading) .section-container {
    min-height: 17.5rem;
  }

  :host(.secondary-heading) .section-container,
  :host(.small) .section-container,
  .section-container {
    flex-direction: column;
    align-items: flex-start;
  }

  .left-container {
    height: fit-content;
  }

  .secondary-container {
    height: fit-content;
    border: none;
  }

  .right-container {
    margin-bottom: 2.5rem;
  }
}

@media (max-width: 480px) {
  :host(.has-background-image) {
    background-image: none !important;
  }

  :host,
  .section-container {
    min-height: 12rem;
  }

  :host,
  .right-container {
    margin-bottom: var(--ic-space-md);
  }

  :host(.small),
  :host(.small) .section-container {
    min-height: 10.813rem;
  }

  :host(.secondary-heading),
  :host(.secondary-heading) .section-container {
    min-height: 16.5rem;
  }

  :host(.secondary-heading) .section-container,
  :host(.small) .section-container,
  .section-container {
    flex-direction: column;
    align-items: flex-start;
  }

  .left-container {
    height: fit-content;
  }

  .secondary-container {
    height: fit-content;
    border: none;
  }
}
