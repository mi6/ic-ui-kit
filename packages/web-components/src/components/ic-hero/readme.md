# ic-hero



<!-- Auto Generated Below -->


## Properties

| Property                    | Attribute                     | Description                                                                       | Type                                 | Default     |
| --------------------------- | ----------------------------- | --------------------------------------------------------------------------------- | ------------------------------------ | ----------- |
| `aligned`                   | `aligned`                     | The alignment of the hero.                                                        | `"center" \| "full-width" \| "left"` | `"left"`    |
| `backgroundImage`           | `background-image`            | The optional background image.                                                    | `string`                             | `undefined` |
| `contentAligned`            | `content-aligned`             | The alignment of the hero content.                                                | `"center" \| "left"`                 | `"left"`    |
| `disableBackgroundParallax` | `disable-background-parallax` | If `true`, the background image (if set) will not scroll using a parallax effect. | `boolean`                            | `false`     |
| `heading` _(required)_      | `heading`                     | The heading of the hero.                                                          | `string`                             | `undefined` |
| `secondaryHeading`          | `secondary-heading`           | The optional secondary heading, replaced by slotted right content.                | `string`                             | `undefined` |
| `secondarySubheading`       | `secondary-subheading`        | The optional secondary subheading, replaced by slotted right content.             | `string`                             | `undefined` |
| `small`                     | `small`                       | If `true`, the small styling will be applied to the hero.                         | `boolean`                            | `false`     |
| `subheading`                | `subheading`                  | The description for the hero.                                                     | `string`                             | `undefined` |


## Slots

| Slot            | Description                                                                   |
| --------------- | ----------------------------------------------------------------------------- |
| `"heading"`     | Content will be rendered in the title area, in place of the heading.          |
| `"interaction"` | Content will be rendered in the interaction area, adjacent to the title area. |
| `"secondary"`   | Content will be rendered in the secondary content area.                       |
| `"subheading"`  | Content will be rendered in the title area, in place of the subheading.       |


## Dependencies

### Depends on

- [ic-section-container](../ic-section-container)
- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-hero --> ic-section-container
  ic-hero --> ic-typography
  style ic-hero fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


