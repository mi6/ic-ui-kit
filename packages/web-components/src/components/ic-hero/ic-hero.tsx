import {
  Component,
  Element,
  Host,
  Prop,
  h,
  State,
  Listen,
} from "@stencil/core";
import {
  IcAlignment,
  IcTheme,
  IcThemeForeground,
  IcThemeForegroundEnum,
} from "../../utils/types";
import {
  slotHasContent,
  getThemeForegroundColor,
  onComponentRequiredPropUndefined,
} from "../../utils/helpers";
import { IcHeroContentAlignments } from "./ic-hero.types";

/**
 * @slot heading - Content will be rendered in the title area, in place of the heading.
 * @slot subheading - Content will be rendered in the title area, in place of the subheading.
 * @slot interaction - Content will be rendered in the interaction area, adjacent to the title area.
 * @slot secondary - Content will be rendered in the secondary content area.
 */

@Component({
  tag: "ic-hero",
  styleUrl: "ic-hero.css",
  shadow: true,
})
export class Hero {
  @Element() el: HTMLIcHeroElement;

  @State() foregroundColor: IcThemeForeground = getThemeForegroundColor();
  @State() rightContent: boolean = false;
  // set by above state
  @State() leftContentFullWidth: boolean =
    !this.rightContent && this.secondaryHeading === undefined;
  @State() scrollFactor: string = "right -100px";

  /**
   * The alignment of the hero.
   */
  @Prop() aligned: IcAlignment = "left";

  /**
   * The optional background image.
   */
  @Prop() backgroundImage?: string;

  /**
   * The alignment of the hero content.
   */
  @Prop() contentAligned: IcHeroContentAlignments = "left";

  /**
   * If `true`, the background image (if set) will not scroll using a parallax effect.
   */
  @Prop() disableBackgroundParallax?: boolean = false;

  /**
   * The heading of the hero.
   */
  @Prop() heading!: string;

  /**
   * The optional secondary heading, replaced by slotted right content.
   */
  @Prop() secondaryHeading?: string;

  /**
   * The optional secondary subheading, replaced by slotted right content.
   */
  @Prop() secondarySubheading?: string;

  /**
   * If `true`, the small styling will be applied to the hero.
   */
  @Prop() small: boolean = false;

  /**
   * The description for the hero.
   */
  @Prop() subheading?: string;

  componentWillLoad(): void {
    this.rightContent = slotHasContent(this.el, "secondary");
  }

  componentDidLoad(): void {
    onComponentRequiredPropUndefined(
      [{ prop: this.heading, propName: "heading" }],
      "Hero"
    );
  }

  componentWillRender(): void {
    this.rightContent = slotHasContent(this.el, "secondary");
  }

  @Listen("themeChange", { target: "document" })
  themeChangeHandler(ev: CustomEvent): void {
    const theme: IcTheme = ev.detail;
    this.foregroundColor = theme.mode;
  }

  @Listen("scroll", { target: "document" })
  doScroll(): void {
    if (this.disableBackgroundParallax) {
      return;
    }

    const scrolltotop = document.scrollingElement.scrollTop;
    const factor = 0.4;
    const y = -100 + scrolltotop * factor;
    this.scrollFactor = "right " + y + "px";
  }

  render() {
    const {
      aligned,
      small,
      heading,
      subheading,
      secondaryHeading,
      secondarySubheading,
      foregroundColor,
      backgroundImage,
      scrollFactor,
    } = this;

    let style = {};

    if (backgroundImage !== undefined) {
      style = {
        "background-image": "url(" + backgroundImage + ")",
        "background-position": scrollFactor,
      };
    }

    return (
      <Host
        class={{
          [IcThemeForegroundEnum.Dark]:
            foregroundColor === IcThemeForegroundEnum.Dark,
          ["has-background-image"]: backgroundImage !== undefined,
          ["small"]: small,
          ["secondary-heading"]: !!secondaryHeading,
        }}
        style={style}
      >
        <ic-section-container
          aligned={aligned}
          fullHeight
          class="section-container"
        >
          <div
            class={{
              ["left-container"]: true,
              ["left-container-full-width"]: this.leftContentFullWidth,
            }}
          >
            <div class="heading">
              <slot name="heading">
                <ic-typography
                  variant={small ? "h2" : "h1"}
                  class={{ ["heading-bottom-spacing"]: !small }}
                >
                  {heading}
                </ic-typography>
              </slot>
            </div>
            <div class="subheading">
              <slot name="subheading">
                <ic-typography variant="body">{subheading}</ic-typography>
              </slot>
            </div>
            <div class="interaction-container">
              <slot name="interaction"></slot>
            </div>
          </div>
          {(secondaryHeading !== undefined || this.rightContent) && (
            <div class="right-container">
              <slot name="secondary">
                {secondaryHeading && (
                  <div class="secondary-container">
                    <div class="secondary-heading">
                      <ic-typography variant="h4">
                        {secondaryHeading}
                      </ic-typography>
                    </div>
                    <div class="secondary-subheading">
                      <ic-typography variant="subtitle-small">
                        {secondarySubheading}
                      </ic-typography>
                    </div>
                  </div>
                )}
              </slot>
            </div>
          )}
        </ic-section-container>
      </Host>
    );
  }
}
