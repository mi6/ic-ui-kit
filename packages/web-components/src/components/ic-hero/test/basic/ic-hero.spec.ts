import { newSpecPage } from "@stencil/core/testing";
import { Hero } from "../../ic-hero";

describe("ic-hero component", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" subheading="Test text"></ic-hero>`,
    });

    expect(page.root).toMatchSnapshot("renders");
  });

  it("should render with a secondary heading", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" subheading="Test text" secondary-heading="Test title" secondary-subheading="Test text"></ic-hero>`,
    });

    expect(page.root).toMatchSnapshot("renders-with-secondary-heading");
  });

  it("should render with interaction content", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" subheading="Test text"><ic-button slot="interaction">Button</ic-button></ic-hero>`,
    });

    expect(page.root).toMatchSnapshot("renders-with-interaction-content");
  });

  it("should render with secondary content", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" subheading="Test text"><ic-card slot="secondary" heading="Test card"></ic-card></ic-hero>`,
    });

    expect(page.root).toMatchSnapshot("renders-with-secondary-content");
  });

  it("should render with a background image when given", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" subheading="Test text" background-image="test.png"><ic-card slot="secondary" heading="Test card"></ic-card></ic-hero>`,
    });

    expect(page.root).toMatchSnapshot("renders-with-background-image");
  });

  it("should use parallax on scroll when a background image is given", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" subheading="Test text" background-image="test.png"><ic-card slot="secondary" heading="Test card"></ic-card></ic-hero>`,
    });

    page.doc.scrollingElement.scrollTop = 50;
    page.doc.dispatchEvent(new CustomEvent("scroll"));

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot("renders-with-parallax-on-scroll");
  });

  it("should use not parallax on scroll when a background image is given, but parallax disabled", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" disable-background-parallax subheading="Test text" background-image="test.png"><ic-card slot="secondary" heading="Test card"></ic-card></ic-hero>`,
    });

    page.doc.scrollingElement.scrollTop = 50;
    page.doc.dispatchEvent(new CustomEvent("scroll"));

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot(
      "renders-with-parallax-on-scroll-disabled"
    );
  });

  it("should render small", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero small heading="Test title" subheading="Test text"></ic-hero>`,
    });

    expect(page.root).toMatchSnapshot("renders-small-variant");
  });

  it("correctly sets foregroundColor on theme change", async () => {
    const page = await newSpecPage({
      components: [Hero],
      html: `<ic-hero heading="Test title" subheading="Test text"></ic-hero>`,
    });

    await page.rootInstance.themeChangeHandler({ detail: { mode: "dark" } });
    await page.waitForChanges();

    expect(page.rootInstance.foregroundColor).toBe("dark");
  });
});
