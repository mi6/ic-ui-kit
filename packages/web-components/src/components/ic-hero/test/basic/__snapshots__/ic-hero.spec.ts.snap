// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-hero component should render small: renders-small-variant 1`] = `
<ic-hero class="small" heading="Test title" small="" subheading="Test text">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container left-container-full-width">
        <div class="heading">
          <slot name="heading">
            <ic-typography variant="h2">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
    </ic-section-container>
  </mock:shadow-root>
</ic-hero>
`;

exports[`ic-hero component should render with a background image when given: renders-with-background-image 1`] = `
<ic-hero background-image="test.png" class="has-background-image" heading="Test title" subheading="Test text" style="background-image: url(test.png); background-position: right -100px;">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container left-container-full-width">
        <div class="heading">
          <slot name="heading">
            <ic-typography class="heading-bottom-spacing" variant="h1">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
      <div class="right-container">
        <slot name="secondary"></slot>
      </div>
    </ic-section-container>
  </mock:shadow-root>
  <ic-card heading="Test card" slot="secondary"></ic-card>
</ic-hero>
`;

exports[`ic-hero component should render with a secondary heading: renders-with-secondary-heading 1`] = `
<ic-hero class="secondary-heading" heading="Test title" secondary-heading="Test title" secondary-subheading="Test text" subheading="Test text">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container">
        <div class="heading">
          <slot name="heading">
            <ic-typography class="heading-bottom-spacing" variant="h1">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
      <div class="right-container">
        <slot name="secondary">
          <div class="secondary-container">
            <div class="secondary-heading">
              <ic-typography variant="h4">
                Test title
              </ic-typography>
            </div>
            <div class="secondary-subheading">
              <ic-typography variant="subtitle-small">
                Test text
              </ic-typography>
            </div>
          </div>
        </slot>
      </div>
    </ic-section-container>
  </mock:shadow-root>
</ic-hero>
`;

exports[`ic-hero component should render with interaction content: renders-with-interaction-content 1`] = `
<ic-hero heading="Test title" subheading="Test text">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container left-container-full-width">
        <div class="heading">
          <slot name="heading">
            <ic-typography class="heading-bottom-spacing" variant="h1">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
    </ic-section-container>
  </mock:shadow-root>
  <ic-button slot="interaction">
    Button
  </ic-button>
</ic-hero>
`;

exports[`ic-hero component should render with secondary content: renders-with-secondary-content 1`] = `
<ic-hero heading="Test title" subheading="Test text">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container left-container-full-width">
        <div class="heading">
          <slot name="heading">
            <ic-typography class="heading-bottom-spacing" variant="h1">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
      <div class="right-container">
        <slot name="secondary"></slot>
      </div>
    </ic-section-container>
  </mock:shadow-root>
  <ic-card heading="Test card" slot="secondary"></ic-card>
</ic-hero>
`;

exports[`ic-hero component should render: renders 1`] = `
<ic-hero heading="Test title" subheading="Test text">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container left-container-full-width">
        <div class="heading">
          <slot name="heading">
            <ic-typography class="heading-bottom-spacing" variant="h1">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
    </ic-section-container>
  </mock:shadow-root>
</ic-hero>
`;

exports[`ic-hero component should use not parallax on scroll when a background image is given, but parallax disabled: renders-with-parallax-on-scroll-disabled 1`] = `
<ic-hero background-image="test.png" class="has-background-image" disable-background-parallax="" heading="Test title" subheading="Test text" style="background-image: url(test.png); background-position: right -100px;">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container left-container-full-width">
        <div class="heading">
          <slot name="heading">
            <ic-typography class="heading-bottom-spacing" variant="h1">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
      <div class="right-container">
        <slot name="secondary"></slot>
      </div>
    </ic-section-container>
  </mock:shadow-root>
  <ic-card heading="Test card" slot="secondary"></ic-card>
</ic-hero>
`;

exports[`ic-hero component should use parallax on scroll when a background image is given: renders-with-parallax-on-scroll 1`] = `
<ic-hero background-image="test.png" class="has-background-image" heading="Test title" subheading="Test text" style="background-image: url(test.png); background-position: right -80px;">
  <mock:shadow-root>
    <ic-section-container aligned="left" class="section-container" fullheight="">
      <div class="left-container left-container-full-width">
        <div class="heading">
          <slot name="heading">
            <ic-typography class="heading-bottom-spacing" variant="h1">
              Test title
            </ic-typography>
          </slot>
        </div>
        <div class="subheading">
          <slot name="subheading">
            <ic-typography variant="body">
              Test text
            </ic-typography>
          </slot>
        </div>
        <div class="interaction-container">
          <slot name="interaction"></slot>
        </div>
      </div>
      <div class="right-container">
        <slot name="secondary"></slot>
      </div>
    </ic-section-container>
  </mock:shadow-root>
  <ic-card heading="Test card" slot="secondary"></ic-card>
</ic-hero>
`;
